medcat/meta_cat.py

```diff
@@ -212,32 +228,52 @@ def train_raw(self, data_loaded: Dict, save_dir_path: Optional[str] = None) -> D
                                  replace_center=g_config['replace_center'], prerequisites=t_config['prerequisites'],
                                  lowercase=g_config['lowercase'])
 
-        # Check is the name there
+        # Check is the name present
         category_name = g_config['category_name']
         if category_name not in data:
             raise Exception(
                 "The category name does not exist in this json file. You've provided '{}', while the possible options are: {}".format(
                     category_name, " | ".join(list(data.keys()))))
 
         data = data[category_name]
+        if data_:
+            data_sampled = []
+            for sample in data_:
+                if isinstance(sample[0][0],str):
+                    doc_text = self.tokenizer(sample[0])
+                    data_sampled.append([doc_text[0]['input_ids'],sample[1],sample[2]])
+                else:
+                    data_sampled.append([sample[0], sample[1], sample[2]])
+            data = data + data_sampled
 
         category_value2id = g_config['category_value2id']
         if not category_value2id:
             # Encode the category values
-            data, category_value2id = encode_category_values(data)
+            data, full_data, category_value2id = encode_category_values(data,category_undersample=self.config.model.category_undersample)
             g_config['category_value2id'] = category_value2id
         else:
             # We already have everything, just get the data
-            data, _ = encode_category_values(data, existing_category_value2id=category_value2id)
-
+            data, full_data, category_value2id = encode_category_values(data, existing_category_value2id=category_value2id,category_undersample=self.config.model.category_undersample)
+            g_config['category_value2id'] = category_value2id
         # Make sure the config number of classes is the same as the one found in the data
         if len(category_value2id) != self.config.model['nclasses']:
             logger.warning(
                 "The number of classes set in the config is not the same as the one found in the data: {} vs {}".format(
                     self.config.model['nclasses'], len(category_value2id)))
             logger.warning("Auto-setting the nclasses value in config and rebuilding the model.")
             self.config.model['nclasses'] = len(category_value2id)
-            self.model = self.get_model(embeddings=self.embeddings)
+
+        self.model = self.get_model(embeddings=self.embeddings)
+
+        if self.config.model.load_model_dict_:
```

Comment:
do we need to be more specific here that we're loading the bert model impl of MetaCAT?

Reply:
we've added the info (in the logger) in the `get_model` function

` def get_model(self, embeddings: Optional[Tensor]) -> nn.Module:
        """Get the model

        Args:
            embeddings (Optional[Tensor]):
                The embedding densor

        Raises:
            ValueError: If the meta model is not LSTM or BERT

        Returns:
            nn.Module:
                The module
        """
        config = self.config
        from medcat.utils.meta_cat.models import LSTM
        from medcat.utils.meta_cat.models import BertForMetaAnnotation
        if config.model['model_name'] == 'lstm':
            model: Union[LSTM, BertForMetaAnnotation] = LSTM(embeddings, config)
            **logger.info("LSTM model used for classification")**

        elif config.model['model_name'] == 'bert':
            model = BertForMetaAnnotation(config)

            if not config.model.model_freeze_layers:

                peft_config = LoraConfig(task_type=TaskType.SEQ_CLS, inference_mode=False, r=8, lora_alpha=16,
                                         target_modules=["query", "value"], lora_dropout=0.2)

                model = get_peft_model(model, peft_config)
                # model.print_trainable_parameters()

            **logger.info("BERT model used for classification")**
`